### DISCUSSION TITLE
How to get serialNumber from a certificate in PEM format .

### DISCUSSION THREAD

**Discussion by CandyMi** [openssl/openssl#24943] [Q&A] [answered]
Hello everyone, here is a pseudo code to demonstrate my question:

```cpp
X509 *x509 = ...;
BIGNUM* bn = ASN1_INTEGER_to_BN(X509_get0_serialNumber(x509), nullptr);
BIO *io = BIO_new(BIO_s_mem());
BN_print(io, bn);
char *buffer;
long length = BIO_get_mem_data(io, &buffer);
dump(buffer, len)
```

  The result is shown below :

```bash
["sn"] = String("7B6B2FA2E0000A3DA1DF710E90A81B512A25082A"),
```

  When I use the `RSA` certificate, the results displayed by the browser are consistent, but when I use the `ECC` certificate, the results displayed by the browser are different.   Is it because of my improper use?

  The test certificate can be found [here](https://gist.github.com/CandyMi/53133fe6345d87d6ee4ad470b9728321).

**Comment by kulkarniamit** [comment, marked as answer]
Are you sure you're downloading the certificate for correct domain name?
For example:
```
$ openssl s_client -connect www.google.com:443 </dev/null 2>&1 | openssl x509 -noout -serial
serial=64A67CADA8DADB210A63E1E8575F7948

$ openssl s_client -connect google.com:443 </dev/null 2>&1 | openssl x509 -noout -serial
serial=718DF8A4D1488A7809CCED27107D8184
```

Here's a sample program that prints the serial number:
```c
#include <stdio.h>
#include <openssl/x509.h>
#include <openssl/bio.h>
#include <openssl/bn.h>
#include <openssl/pem.h>

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "[ERR] Please provide a certificate input\n");
        return 1;
    }
    const char *cert_file = argv[1];
    FILE *fp = fopen(cert_file, "r");
    if (!fp) {
        /* handle error and return */
    }

    X509 *x509 = PEM_read_X509(fp, NULL, NULL, NULL);
    fclose(fp);
    if (!x509) {
        /* handle error and return */
    }

    const ASN1_INTEGER *serial = X509_get0_serialNumber(x509);
    BIGNUM *bn = ASN1_INTEGER_to_BN(serial, NULL);
    if (!bn) {
        /* handle error, release memory and return */
    }

    BIO *io = BIO_new(BIO_s_mem());
    if (!io) {
        /* handle error, release memory and return */
    }

    /* Print the BIGNUM to the BIO */
    if (BN_print(io, bn) <= 0) {
        /* handle error, release memory and return */
    }

    /* Print the data from BIO */
    BIO *out = BIO_new_fp(stdout, BIO_NOCLOSE);
    char *buffer;
    long length = BIO_get_mem_data(io, &buffer);
    buffer[length] = '\0';
    BIO_printf(out, "serial=%s\n", buffer);

    BIO_free(out);
    BIO_free(io);
    BN_free(bn);
    X509_free(x509);

    return 0;
}
```

### Test with certificates with RSA and EC keys
* RSA
  ```bash
  $ openssl s_client -connect www.example.org:443  </dev/null 2>&1 | \
  sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' | \
  openssl x509 -noout -text |grep -A1 -i "Public key alg\|serial"
          Serial Number:
              07:5b:ce:f3:06:89:c8:ad:df:13:e5:1a:f4:af:e1:87
  --
              Public Key Algorithm: rsaEncryption
                  Public-Key: (2048 bit)

  $ openssl s_client -connect www.example.org:443  </dev/null 2>&1| \
  sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > cert-for-rsa.pem
  
  $ ./print_cert_serial cert-for-rsa.pem
  serial=75BCEF30689C8ADDF13E51AF4AFE187

  $ openssl x509 -in cert-for-rsa.pem -noout -serial
  serial=075BCEF30689C8ADDF13E51AF4AFE187  
  ```

* ECC
  ```bash
  $ openssl s_client -connect www.google.com:443 </dev/null 2>&1 | openssl x509 -noout -text | \
  grep -i -A1 "public key alg\|serial"
          Serial Number:
              64:a6:7c:ad:a8:da:db:21:0a:63:e1:e8:57:5f:79:48
  --
              Public Key Algorithm: id-ecPublicKey
                  Public-Key: (256 bit)
  
  $ openssl s_client -connect www.google.com:443  </dev/null 2>&1 | \
  sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > cert-for-ecc.pem
    
  $ ./print_cert_serial cert-for-ecc.pem
  serial=64A67CADA8DADB210A63E1E8575F7948

  $ openssl x509 -in cert-for-ecc.pem -noout -serial
  serial=64A67CADA8DADB210A63E1E8575F7948
  ```

**Reply by CandyMi** [reply]
Thank you for your answer! @kulkarniamit 

After my test, the results of the OpenSSL command line are consistent with my code.

However, the results displayed by the Edge browser are different, and I don’t know why.

```bash
candymi@localhost:~$ openssl s_client -connect localhost:8080 </dev/null 2>&1 | openssl x509 -noout -serial
serial=93FC718E7FFB6FC9
```

![image](https://github.com/user-attachments/assets/a380347b-24da-405a-aaae-4cf69d5b4335)

**Reply by kulkarniamit** [reply]
Sorry I might be missing something but I see the same result in the browser cert as shown by your `openssl s_client` command: `93FC718E7FFB6FC9`.
The browser is displaying the same serial hexadecimal characters separated by colon (`:`).
Can you elaborate on what's different in the browser?

**Reply by CandyMi** [reply]
What I mean is:

* When the certificate is RSA, the browser and command line output are consistent.

* When the certificate is ECC, the browser will include the "00" prefix, but openssl does not.

**Reply by kulkarniamit** [reply]
`serialNumber` field in a certificate is of type `CertificateSerialNumber` ([Reference](https://datatracker.ietf.org/doc/html/rfc5280#page-117)) and that's an `INTEGER`.
If MSB of that integer is 1, the integer might be negative (two's complement representation), a leading zero byte might be required to avoid ambiguity. If it's `0`, padding byte `0x00` is not required.

* Serial without leading `0x00` byte (msb `0`)
```bash
# Create a self signed certificate with a serial with MSB 0 (Example: 7B6B2FA2E0000A3DA1DF710E90A81B512A25082A)
$ openssl req -x509 -newkey rsa:2048 -keyout customserial-rsa-key.pem -out customserial-rsa-cert.pem -days 365 \
-nodes -set_serial 0x7B6B2FA2E0000A3DA1DF710E90A81B512A25082A  \
-subj "/C=CN/ST=Beijing/L=Beijing/O=Random Company Ltd./OU=IT Department/CN=example.com" > /dev/null 2>&1

# Check the offset where serial can be found in the certificate
$ openssl asn1parse -in customserial-rsa-cert.pem |grep "7B6B2FA2E0000A3DA1DF710E90A81B512A25082A"
   13:d=2  hl=2 l=  20 prim: INTEGER           :7B6B2FA2E0000A3DA1DF710E90A81B512A25082A

# As seen above, offset is 13 and length is 20, print those bytes
# There should be no leading 0x00 byte
$ openssl x509 -in customserial-rsa-cert.pem -outform DER |xxd --seek 13 -l $((2+20)) -p
02147b6b2fa2e0000a3da1df710e90a81b512a25082a

# 0x02 is the type (INTEGER)
# 0x14 is the upcoming length (20 bytes)
```

* Serial with leading `0x00` byte (msb `1`)
```bash
# Create a self signed certificate with a serial with MSB 1 (Example: 93FC718E7FFB6FC9)
$ openssl req -x509 -newkey rsa:2048 -keyout customserial-ecc-key.pem -out customserial-ecc-cert.pem -days 365 \
-nodes -set_serial 0x93FC718E7FFB6FC9  \
-subj "/C=CN/ST=Beijing/L=Beijing/O=Random Company Ltd./OU=IT Department/CN=example.com" > /dev/null 2>&1

# Check the offset where serial can be found in the certificate
$ openssl asn1parse -in customserial-ecc-cert.pem |grep "93FC718E7FFB6FC9"
   13:d=2  hl=2 l=   9 prim: INTEGER           :93FC718E7FFB6FC9

# As seen above, offset is 13 and length is 9, print those bytes
# There should be a leading 0x00 byte
$ openssl x509 -in customserial-ecc-cert.pem -outform DER |xxd --seek 13 -l $((2+9)) -p
02090093fc718e7ffb6fc9

# 0x02 is the type (INTEGER)
# 0x09 is the upcoming length (9 bytes including leading 0x00)
```

### Links for more information
* [Certificate ASN1 format](https://datatracker.ietf.org/doc/html/rfc5280#page-117)
* [Integer padding](https://coolaj86.com/articles/asn1-for-dummies/)
* [Tag values for different data types in ASN1](https://letsencrypt.org/docs/a-warm-welcome-to-asn1-and-der/#tag)
* [BER encoding section has some good examples](https://luca.ntop.org/Teaching/Appunti/asn1.html)
* [Decode cert with ECC and leading MSB 1](https://lapo.it/asn1js/#MIIBjDCCATMCCQCT_HGOf_tvyTAJBgcqhkjOPQQBME8xCzAJBgNVBAYTAmNuMQswCQYDVQQIDAJjZDEVMBMGA1UEBwwMRGVmYXVsdCBDaXR5MRwwGgYDVQQKDBNEZWZhdWx0IENvbXBhbnkgTHRkMB4XDTE5MDgyMzA3MzEyNVoXDTIwMDgyMjA3MzEyNVowTzELMAkGA1UEBhMCY24xCzAJBgNVBAgMAmNkMRUwEwYDVQQHDAxEZWZhdWx0IENpdHkxHDAaBgNVBAoME0RlZmF1bHQgQ29tcGFueSBMdGQwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASHWxVhsXJfEjojWhSGl-a-g5udAc3Uq_46nVNxGilEb1LshvS5kezj9FPimeFXeJXoXrae3TyEtqP-ukJIL822MAkGByqGSM49BAEDSAAwRQIgTdsjH7u2hvEVFTYh3kdaUjMmcNpHiCwcxegPHAVS1JkCIQDhrtwjCHsamSk0alrqyb7bzAoZyg7ZBDa4hvT9ELoThw)
* [Decode cert with RSA and MSB 0](https://lapo.it/asn1js/#MIIDATCCAekCFHtrL6LgAAo9od9xDpCoG1EqJQgqMA0GCSqGSIb3DQEBCwUAMD0xCzAJBgNVBAYTAkNOMQswCQYDVQQIDAJDRDEhMB8GA1UECgwYSW50ZXJuZXQgV2lkZ2l0cyBQdHkgTHRkMB4XDTE5MTExMjAyMjQ0NVoXDTIwMTExMTAyMjQ0NVowPTELMAkGA1UEBhMCQ04xCzAJBgNVBAgMAkNEMSEwHwYDVQQKDBhJbnRlcm5ldCBXaWRnaXRzIFB0eSBMdGQwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCsQ-jcs33WqhynMDJl6tMbwhvobEN3zl65TcWII-chetPwN9xnfNfGDfpu8F_8w47ONRRwLDCNmM_8tSDJiqtaLaxZF6wHdiWBQR55jUHSwVoy6eVAmOHgArKCloj3xK7HkQz668SalZ1LSePJeIlZycimgFrnSzsP6jbdtA9cMHuMWGBCVb23C8A5-L2C_zO1bo22djrtxsm_TiBM39WplXc6OLfaPPI5THDA_EnSyvv7Wur9WSmT_ZZPl0aEPtYwoYLLSDYgtmjx-7tjqLe-jHZdWUzknaEPFlQU3wlVdTMjbCu7pi0tBp8fk-G4EQX1Vr4WbkKHAs_UeF95s15HAgMBAAEwDQYJKoZIhvcNAQELBQADggEBAB_MLRHcW2En94QxnpTZjJ7c05SOgaYXE_E2OEuQeqbekfqdkpxV6YV9CURhYGdbhPh6CO6Z8wy1d1jN76dNlcKXxxZ9qM2IRllsfJkw_Ao_5eSYX_gxBJx51Yxxj_itwqtrROk5nlyh5hNa2y-2ANyqJmBDtKU93qHCAeyIDEo5rPqfvazflwfGDu-9zmUUeyMoNQxXMINgbtJbNQUN74uZXxBoFQiYmfQEEZpr6NNOFkFHXwsE6fsu97SIyz8PLB7o9C3JK3YhbnnzforZHIy4STtfWoXAo43uZa0ufRnuVJvXQgjUkBgSqdS8xVEyAbuOJB_w5TBeo6BeRd1_wJ0)

**Reply by CandyMi** [reply]
Got it, thank you for your detailed answer.
